File: Python/Medium/515.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def largestValues(self, root: Optional[TreeNode]) -> List[int]:
        # bfs

        # empty tree
        if root is None:
            return list()

        answer = list()
        answer.append(root.val)

        q = list()
        q.append(root)

        while len(q):
            max = -(2**31)

            qSize = len(q)
            for i in range(qSize):
                holder = q.pop(0)

                if holder.left:
                    q.append(holder.left)
                    max = max if (max > holder.left.val) else holder.left.val

                if holder.right:
                    q.append(holder.right)
                    max = max if (max > holder.right.val) else holder.right.val

            if len(q):
                answer.append(max)

        return answer

File: Java/Medium/515.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 * int val;
 * TreeNode left;
 * TreeNode right;
 * TreeNode() {}
 * TreeNode(int val) { this.val = val; }
 * TreeNode(int val, TreeNode left, TreeNode right) {
 * this.val = val;
 * this.left = left;
 * this.right = right;
 * }
 * }
 */
class Solution {
    public List<Integer> largestValues(TreeNode root) {
        // bfs
        if (root == null) {
            return new ArrayList<Integer>();
        }

        ArrayList<Integer> answer = new ArrayList<Integer>();
        ArrayList<TreeNode> q = new ArrayList<TreeNode>();

        q.add(root);
        answer.add(root.val);

        int size, max;
        TreeNode holder;

        while (q.size() > 0) {

            size = q.size();
            max = Integer.MIN_VALUE;

            for (int i = 0; i < size; i++) {
                holder = q.remove(0);

                if (holder.left != null) {
                    q.add(holder.left);
                    max = (max > holder.left.val) ? max : holder.left.val;
                }

                if (holder.right != null) {
                    q.add(holder.right);
                    max = (max > holder.right.val) ? max : holder.right.val;
                }

            }

            if (q.size() > 0) {
                answer.add(max);
            }
        }

        return answer;

    }
}